module/zfs/dsl_dir.c

```diff
@@ -1803,6 +1811,21 @@ dsl_dir_rename_check(void *arg, dmu_tx_t *tx)
 				dsl_dir_rele(dd, FTAG);
 				return (err);
 			}
+
+			/*
+			 * encrypted datasets can only be moved if they are
+			 * an encryption root (locally set keyformat).
```

Comment:
> Why don't we allow that? Seems like a common use case would be "encrypt all datasets below this the same way". "whoops, named a child dataset wrong, rename it"

To clarify, renaming an encryption child is ok. Moving to a new parent is not currently.

> As long as the new location is under the same encryption root, I think we can allow the rename. Would that be hard to implement?

I think I can implement it like that. Shouldn't be a big deal.

> I thought there was a restriction that an unencrypted dataset couldn't be under an encrypted dataset. Doesn't that need to be enforced here?

Whoops. It looks like I must've rebased that away at some point. I will put it back.

> I didn't see either of these restrictions in the manpages. Seems like they should at least be in the zfs rename section, and maybe also the Encryption section.

I will make sure this is documented in both places (I think there is a sentence about it in the encryption section).